message GameMessage {
  enum Type {
    BALL_POSITIONS = 1;
    SCORE = 2;
    SERVER_RELEASE_CONTROL = 3;
    CLIENT_HIT = 4;
  }

  required Type type = 1;

  optional BallPositions ball_positions = 2;
  optional Score score = 3;
  optional HitInfo client_hit = 4;
}

message Vector3f {
  required float x = 1;
  required float y = 2;
  required float z = 3;
}

message BallPositions {
  repeated Vector3f ball = 1;
}

message Score {
  required float host_score = 1;
  required float client_score = 2;
}

message HitInfo {
  enum Strength { LOW = 1; MEDIUM = 2; HIGH = 3; }
  required Strength strength = 1;
  required Vector3f direction = 2;
}

